### DISCUSSION TITLE
Question on documentations

### DISCUSSION THREAD

**Discussion by Alex-888888** [jankae/LibreVNA#241] [General]
Dear jankae,

Thank you for libreVNA and all the shared informations.
Inside DevelopperInfo section are two documents : Device_protocol_v13.pdf and USB_protocol_v12.pdf.
They are different and some USB packet (for spectrum analyser for example) are different.
Which is the reference document ?
I think Device_protocol_v13.pdf is the good one but before starting any coding, I would like to be sure.
Inside Device_protocol_v13.pdf, an Ethernet interface is described.
Is it TCP over USB or it's possible to add a physical Ethernet ?
Thank you for your answer.
Best regards,
Alex

**Comment by jankae** [comment]
Hi Alex,
version 13 is the one used in the latest firmware, see also here: https://github.com/jankae/LibreVNA/blob/2bbb8792d83e8c3548544114d37c3205a27452f5/Software/VNA_embedded/Application/Communication/Protocol.hpp#L13

Version 12 was in use in an older firmware (and versions 1-11 have also been used in even older firmwares but they are not documented).

> Inside Device_protocol_v13.pdf, an Ethernet interface is described.
Is it TCP over USB or it's possible to add a physical Ethernet ?

It is possible to add physical Ethernet (the GUI fully supports it already) but that has only been implemented on a non-published prototype of mine. If you want to go that route, you are on your own - at least for now.

**Reply by Alex-888888** [reply]
Dear Jan,

Thanks for your answer.
If it's possible for you to send more information on how to add a physical Ethernet, it would be nice. 
Protocol of Communication is the same ?
Best regards,
Alex

**Reply by jankae** [reply]
The protocol is the same. There are some additional things that happen with Ethernet (e.g. discovery of LibreVNA over the SSDP protocol) but this is all described in the documentation for protocol v13 (section 2.2).

Adding Ethernet to the LibreVNA requires significant changes. You would need to use a different µC (or add an Ethernet MAC) and also change a lot in the firmware (at least a lot of the low-level stuff, once you get to the point where your Ethernet-Stack sends and receives packets according to the protocol description the high-level stuff can stay the same).

I am sorry but I can't share any details of my implementation, it was a customer project and I was just re-using some ideas from the LibreVNA and thought that adding TCP support to the GUI at that point would be a good idea for the future.

The general idea for implementing Ethernet is this:

- Pick a different STM32 with an Ethernet controller peripheral
- When you set up the project in CubeMX, enable the Ethernet and lwip
- Configure lwip in the code to listen at the correct ports (described in the protocol description document)
- Throw received data into the packet parsing functions (the same that are already used by the USB stack)
- Send encoded packets through the lwip socket instead of through USB

Depending on your level of experience this might either be an easy task or a completely impossible project. The good news is that you "only" need to change the hardware and firmware, the GUI software is already handling its side of the communication.

**Reply by Alex-888888** [reply]
Dear Jan,

Thanks for all the shared informations.
I will think about all.
Best regards,

Alex

**Comment by Alex-888888** [comment]
Dear Jan,

One last question.
Is it possible to reduce the number of point for the sweep in SA mode trough the SCPI server ?
Best regards,

Alex

**Reply by jankae** [reply]
No, changing the number of points in SA mode is not possible at the moment. It shouldn't be too hard to implement but if you are trying to achieve faster sweep speeds with that (by reducing the number of points) it will not work: The measurement process of an SA is different from a VNA, it does not know which frequencies to look for. That means it needs a certain frequency spacing in order to not miss any peaks (this spacing depends on the RBW). The LibreVNA calculates the required spacing internally and will measure the required amount of points regardless of the number of requested points. It will then reduce the measured points to the number of requested points per sweep according to the selected detector type.
This means that reducing the number of requested points will not get you a faster sweep unless your span to RBW ratio is in the same order of magnitude as the amount of requested points (meaning a very narrow span).
On the other hand, if you are after more than the standard 1001 points this could be a useful feature.

**Reply by Alex-888888** [reply]
Jan,

Thanks for the explainations.
If my memory works fine, I remember that in SA you could also play with the video bandwidth.
I want to check narrow bandwidth and only increasinf RBW gives sometimes strange results.
If I want to modify the number of points in SA were should I give a look.
GUI software, µC or FPGA.
Best regards,
Alex

**Reply by jankae** [reply]
> If my memory works fine, I remember that in SA you could also play with the video bandwidth.

Yes, that is a feature which is not available in the LibreVNA. I guess you could emulate it by smoothing the trace with a median filter (which is available as a math operation).

> I want to check narrow bandwidth and only increasinf RBW gives sometimes strange results.

That is something you can do already now. Narrow bandwidths will significantly increase the sweep time but that is true for all spectrum analyzers and reducing the numbers of displayed points will not help.

> If I want to modify the number of points in SA were should I give a look.
GUI software, µC or FPGA.

Changing it in the GUI should be enough if I remember correctly.

**Reply by Alex-888888** [reply]
Jan,

Thank you for your reply.
Which development tools are you using ?
Your job on the libre VNA is incredible.

**Reply by jankae** [reply]
You should find all the relevant information here: https://github.com/jankae/LibreVNA/blob/master/Documentation/DeveloperInfo/BuildAndFlash.md
Please let me know if something is missing there and I will add it.

> Your job on the libre VNA is incredible.

Thank you :) it is certainly not free from errors but I try my best to improve it

**Reply by Alex-888888** [reply]
Thanks, it's fine.
I will learn new thing because I've never used Qt and I'm under Windows...
I will give you a feddback when I will be able to compile it fine on my computer. ;)

**Reply by Alex-888888** [reply]
Jan,
I was able to rebuild the libreVNA GUI project under windows 11.
Qt creator installation didn't gave any problem.
The libusb-1.0 installation under windows 11 should be done manually and the .a files should be placed directly inside the folder of the LibreVNA-GUI.pro.
As I don't want the .a files inside this folder I have modify the LibreVNA-GUI.pro as follow :
```
#win32:LIBS += -L"$$_PRO_FILE_PWD_" # Github actions placed libusb here
win32:LIBS += -LC:\Qt\6.6.2\mingw_64\lib
```
If it can help.

I will now give alook to the µC. :)

**Reply by Alex-888888** [reply]
I think I was able to rebuild to µC also but I face some troubles.
Do you have the last vna_embeded.bin release in order to compare my file and yours ?
Thanks,
Alex

**Reply by jankae** [reply]
You can find the VNA_embedded.elf (not bin, but I think this should also work for comparison) in the firmware files for each commit. E.g. here from the latest one: https://github.com/jankae/LibreVNA/actions/runs/7948406082

**Reply by Alex-888888** [reply]
Thanks,
Compilation is not correctly working on windows for the moment.
I have the following warnings and error...
```
dwarf line number information for .data ignored
leb128 operand is an undefined symbol: .LVU16
```
I will go to linux to try to rebuild µC

**Reply by Alex-888888** [reply]
Dear Jan,
It's also impossible to build the uC from linux with the latest release of CubeIde.
Which release have you used to build the project?
Alex

**Reply by jankae** [reply]
I am still using 1.8.0. It is a bit old but works fine for me. I just tried it with 1.14.1 and get the same error as you. Some googling shows that it is likely a binutils bug(?), it doesn't seem to like executable code in the .data section combined with compiling for debug.

I need the executable .data for the firmware update, so the only option is to disable the `-g` option in the compiler:
![image](https://github.com/jankae/LibreVNA/assets/16818866/fb79e357-3be0-47b2-a7e4-e52205f75127)

Once I do that, I can build it with 1.14.1 as well. I guess this would make debugging a bit more difficult, so I will be staying on 1.8.0 for now.

**Reply by Alex-888888** [reply]
Jan,
Thanks for your help.
In this way I can also build the project.
Googling never point me to a binutils bug. 
I spent some days trying to rebuild the project from many ways.
I without your answer i was in the point to recreate project from scratch.
I will continue to search to find a way to enable the -g option in the compiler.
Thank you
